misc/python/materialize/mzexplore/catalog/u_clone_dependencies.sql

```diff
@@ -0,0 +1,76 @@
+-- Copyright Materialize, Inc. and contributors. All rights reserved.
+--
+-- Use of this software is governed by the Business Source License
+-- included in the LICENSE file at the root of this repository.
+--
+-- As of the Change Date specified in that file, in accordance with
+-- the Business Source License, use of this software will be governed
+-- by the Apache License, Version 2.0.
+
+-- Starting from a set of `source_ids` that share the same `cluster_id`, find
+-- all catalog objects that need to be cloned in order to perform a runtime
+-- experiment to asses the impact plan chages to the source_ids. 
+--
+-- The returned set of objects is the union of these cases:
+-- 1. The original set of source_ids themselves.
+-- 2. All source_ids representing views that transitively depend on source_ids.
+-- 3. All indexes on objects adjacent to (2) that are in the same cluster.
+WITH
+  -- Pre-define all direct dependencies on views.
+  direct_view_dependencies(source_id, target_id) AS (
+    SELECT
+      od.object_id, od.referenced_object_id
+    FROM
+      mz_internal.mz_object_dependencies od JOIN
+      mz_catalog.mz_objects o ON(od.object_id = o.id)
+    WHERE
+      od.referenced_object_id ilike 'u%' AND o.type = 'view'
+  ),
+  -- Find the transitive dependencies reachable along view paths for all source_ids.
+  transitive_dependencies(source_id, target_id) AS (
+    WITH MUTUALLY RECURSIVE
+      reach(source_id text, target_id text) AS (
+        -- Start from requested source_ids.
+        SELECT object_id, referenced_object_id FROM mz_internal.mz_object_dependencies WHERE object_id IN ({source_ids})
+
+        UNION
+
+        -- In each iteration, enrich with adjacent direct_view_dependencies.
+        SELECT x, z FROM reach r(x, y) JOIN direct_view_dependencies d(y, z) USING(y)
+      )
+    SELECT source_id, target_id FROM reach
+  ),
+  -- Find indexes on transitive_dependencies targets that:
+  -- 1. would be available for at least one of the requested source_ids, and
+  -- 2. live in the requested cluster.
+  dependency_indexes(id) AS (
+    SELECT
+      id
+    FROM
+      mz_catalog.mz_indexes i JOIN
+      transitive_dependencies d ON (i.on_id = d.target_id)
+    WHERE
+      id <= (SELECT max(id) FROM unnest(array[{source_ids}]) AS src(id)) AND -- (1)
```

Comment:
> I'm pretty sure the max uses lexicographical order, which means u100 will be smaller than u9, so it doesn't reflect creation order.

Good catch.

> I'm less sure about that, but when envd restarts, does it take into account creation order when choosing indexes for dataflows?

This is my assumption—I was thinking that if an index is created _after_ a dataflow by the user will also always be created after the dataflow upon restart. We also depend on `EXPLAIN REPLAN` for that, but I agree that it's not the most robust solution.

> Btw, if you are interested in the set of indexes actually used by the currently installed dataflows, you can find these in mz_internal.mz_compute_dependencies.

This might be more robust, but possibly also too restrictive. We want to export indexes that the alternative planning strategy will select, which is a subset of the indexes made available to the `IndexOracle` during replanning. For example, assume that I want to clone a dataflow definition `mv` which originally would see three indexes `i1`, `i2`, `i3` on a view `v` referenced in the `mv` definition. The current `mv` plan might only use `i2`, but I might want to test an alternative optimizer pipeline will use both `i1` and `i2`.

Reply:
> I was thinking that if an index is created after a dataflow by the user will also always be created after the dataflow upon restart.

I tested locally with this simple SQL and it looks like you are correct.

```sql
CREATE TABLE t (a int);
CREATE MATERIALIZED VIEW mv AS SELECT * FROM t;
CREATE DEFAULT INDEX ON t;
```

I don't remember seeing anything in the bootstrap code that ensures this property though. Might be worth confirming with the adapter team whether this is a guarantee they provide or just an implementation detail that can accidentally change at any time!

> This might be more robust, but possibly also too restrictive.

This makes sense to me.

Reply:
> Might be worth confirming with the adapter team whether this is a guarantee they provide or just an implementation detail that can accidentally change at any time!

I did confirm with @sploiselle last week, but now that I think about it I will just add an `*.slt` test with something along the above and a big fat comment that explains where we currently rely on that behavior.